wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.kt

```diff
@@ -0,0 +1,161 @@
+import java.util.LinkedHashMap
+
+internal class MessageTypeAdapter<M : Message<M, B>, B : Message.Builder<M, B>>(
+  private val gson: Gson,
+  type: TypeToken<M>
+) : TypeAdapter<M>() {
+  private val messageAdapter: RuntimeMessageAdapter<M, B> =
+      RuntimeMessageAdapter.create(type.rawType as Class<M>)
+  private val fieldBindings: Map<String, FieldBinding<M, B>>
+
+  init {
+    val fieldBindings = LinkedHashMap<String, FieldBinding<M, B>>()
+    for (binding in messageAdapter.fieldBindings().values) {
+      fieldBindings[binding.name] = binding
+    }
+    this.fieldBindings = unmodifiableMap(fieldBindings)
+  }
+
+  @Throws(IOException::class)
+  override fun write(out: JsonWriter, message: M?) {
+    if (message == null) {
+      out.nullValue()
+      return
+    }
+
+    out.beginObject()
+    for (tagBinding in messageAdapter.fieldBindings().values) {
+      val value = tagBinding.get(message) ?: continue
+      out.name(tagBinding.name)
+      emitJson(out, value, tagBinding.singleAdapter(), tagBinding.label)
+    }
+    out.endObject()
+  }
+
+  private fun emitJson(out: JsonWriter, value: Any, adapter: ProtoAdapter<*>, label: Label) {
+    if (adapter === ProtoAdapter.UINT64) {
+      if (label.isRepeated) {
+        val longs = value as List<Long>
+        out.beginArray()
+        val count = longs.size
+        for (i in 0 until count) {
+          emitUint64(longs[i], out)
+        }
+        out.endArray()
+      } else {
+        emitUint64(value as Long, out)
+      }
+    } else {
+      gson.toJson(value, value.javaClass, out)
+    }
+  }
+
+  private fun emitUint64(value: Long, out: JsonWriter) {
+    if (value < 0) {
+      val unsigned = POWER_64.add(BigInteger.valueOf(value))
+      out.value(unsigned)
+    } else {
+      out.value(value)
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun read(`in`: JsonReader): M? {
+    if (`in`.peek() == JsonToken.NULL) {
+      `in`.nextNull()
+      return null
+    }
+
+    val elementAdapter = gson.getAdapter(JsonElement::class.java)
+    val builder = messageAdapter.newBuilder()
+
+    `in`.beginObject()
+    while (`in`.peek() != JsonToken.END_OBJECT) {
+      val name = `in`.nextName()
+
+      val fieldBinding = fieldBindings[name]
+      if (fieldBinding == null) {
+        `in`.skipValue()
+      } else {
+        val element = elementAdapter.read(`in`)
+        val value = parseValue(fieldBinding, element)
+        fieldBinding.set(builder, value)
+      }
+    }
+
+    `in`.endObject()
+    return builder.build()
+  }
+
+  private fun parseValue(fieldBinding: FieldBinding<*, *>, element: JsonElement): Any {
+    if (fieldBinding.label.isRepeated) {
+      if (element.isJsonNull) {
+        return emptyList<Any>()
+      }
+
+      val itemType = fieldBinding.singleAdapter().javaType
+
+      val array = element.asJsonArray
+      val result = ArrayList<Any>(array.size())
+      for (item in array) {
+        result.add(gson.fromJson(item, itemType))
```

Comment:
`return element.asJsonArray.map { gson.fromJson(it, itemtype)}` ?

Reply:
Fixed